ExampleMod/Projectiles/ExampleLastPrismBeam.cs

```diff
@@ -0,0 +1,353 @@
+			Vector2 samplingPoint = projectile.Center;
+			if (fullCharge)
+				samplingPoint = prism.Center;
+			if (!Collision.CanHitLine(Main.player[projectile.owner].Center, 0, 0, prism.Center, 0, 0))
+				samplingPoint = Main.player[projectile.owner].Center;
+
+			// Perform a laser scan to calculate the correct length of the beam.
+			// Alternatively, if you want the beam to ignore tiles, just set it to be the max beam length with the following line.
+			// BeamLength = MaxBeamLength;
+
+			float[] laserScanResults = new float[NumSamplePoints];
+			Collision.LaserScan(samplingPoint, projectile.velocity, BeamTileCollisionWidth * projectile.scale, MaxBeamLength, laserScanResults);
+			float averageLengthSample = 0f;
+			for (int i = 0; i < laserScanResults.Length; ++i)
+				averageLengthSample += laserScanResults[i];
+			averageLengthSample /= NumSamplePoints;
+
+			return averageLengthSample;
+		}
+
+		// Determines whether the specified target hitbox is intersecting with the beam.
+		public override bool? Colliding(Rectangle projHitbox, Rectangle targetHitbox)
+		{
+			// If the target is touching the beam's hitbox (which is a small rectangle vaguely overlapping the host Prism), that's good enough.
+			if (projHitbox.Intersects(targetHitbox))
+				return true;
+			// Otherwise, perform an AABB line collision check to check the whole beam.
+			float _ = float.NaN;
+			Vector2 beamEndPos = projectile.Center + projectile.velocity * BeamLength;
+			return Collision.CheckAABBvLineCollision(targetHitbox.TopLeft(), targetHitbox.Size(), projectile.Center, beamEndPos, BeamHitboxCollisionWidth * projectile.scale, ref _);
+		}
+
+		public override bool PreDraw(SpriteBatch spriteBatch, Color lightColor)
+		{
+			// If the beam doesn't have a defined direction, don't draw anything.
+			if (projectile.velocity == Vector2.Zero)
+				return false;
+
+			Texture2D tex = Main.projectileTexture[projectile.type];
+			Vector2 centerFloored = projectile.Center.Floor() + projectile.velocity * projectile.scale * 10.5f;
+			Vector2 scaleVec = new Vector2(projectile.scale);
+
+			// Reduce the beam length proportional to its square area to reduce block penetration.
+			float visualBeamLength = BeamLength - 14.5f * projectile.scale * projectile.scale;
+
+			DelegateMethods.f_1 = 1f; // f_1 is an unnamed decompiled variable whose function is unknown. Leave it at 1.
+			Vector2 beamStartPos = centerFloored - Main.screenPosition;
+			Vector2 beamEndPos = beamStartPos + projectile.velocity * visualBeamLength;
+			Utils.LaserLineFraming llf = new Utils.LaserLineFraming(DelegateMethods.RainbowLaserDraw);
+
+			// Draw the outer beam.
+			// c_1 is an unnamed decompiled variable which is the render color of the beam drawn by DelegateMethods.RainbowLaserDraw.
+			Color outerBeamColor = GetBeamColor();
+			DelegateMethods.c_1 = outerBeamColor * OuterBeamOpacityMultiplier * projectile.Opacity;
+			Utils.DrawLaser(Main.spriteBatch, tex, beamStartPos, beamEndPos, scaleVec, llf);
```

Comment:
use just `spriteBatch`, not `Main.spriteBatch` (even tho its the same object 99% of the time, just better using the passed one instead of Main.)

Reply:
I forgot that there was a passed in sprite batch. Whoops.